### PR TITLE
Addons, core: Make `react` and Storybook packages `devDependencies` where possible

### PR DESCRIPTION
**Pull request by JReinhold**
Works on #24490 

## ~TODO~
- [ ] ~Document new `optimizeDeps` preset property~
- [x] Reword "no React" error
- [x] Test that an addon author can import from `manager-api` and still have correct types, even though `@storybook/types` and `@types/react` are not part of the package.

## What I did

This PR is a retry of #23486 with many non-essential things ignored.

The main purpose of this PR is to minimize the amount of peer dependencies in addons and a few of the core packages, most notably removing `react` as a peer dependency.

The way the manager and the preview globalizes a fixed set of dependencies means that these packages don't strictly need the dependencies, as they will be referenced from the global window instead, which our builders ensures. Removing them from peer dependencies mean that users don't need to specify them either.

Having them as devDependencies will however make our (tsup) bundler default to bundle them in, which we don't want. Therefore this PR also introduces a new bundler configuration, `addon-bundle` that makes sure that the globalized packages are externalized (thus the imports are kept intact), so they can be replaced with the global reference when built.
A handful of packages like `manager-api` are _not_ using `addon-bundle` but still need these packages externalized, so they've been added to their bundler configuration manually to ensure that happens.

`addon-essentials` and `addon-docs` are largely left out of this PR, as tackling them in terms of `react` requires more work.

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [x] end-to-end tests

#### Manual testing

It's important to test this in the different package manager configuration, npm, yarn, yarn pnp and pnpm and with both builders.

For each configuration and builder:

1. Create a new non-React sandbox outside of the monorepo (`npx sb@next sandbox`)
2. Replace all versions with this PR's canary version.

#### Testing `addon-essentials` _with_ docs

3. Run `storybook` and `build-storybook` and see that it works

#### Testing `addon-docs` directly

1. Replace `addon-essentials` with `addon-docs` both in `main.ts` and `package.json`
2. Run `storybook` and `build-storybook` and see that it works

#### Testing `addon-essentials` with docs _disabled_

1. Remove the `mdx` glob from `stories` in `main.ts`
1. Replace `addon-docs` with `addon-essentails` both in `main.ts` and `package.json` again
2. Remove `react` and `react-dom` from `package.json`. You will see an unmet peer deps from `addon-essentials`, that's okay.
3. disable `docs` in `main.ts` with `{ name: '@storybook/addon-essential', options: { docs: false } }`
4. Run `storybook` and `build-storybook` and see that it works, now without React

#### Testing _without_ `addon-essentials` and `addon-docs`

1. Remove `addon-docs` with `addon-essentails` both from `main.ts` and `package.json` again
2. Run `storybook` and `build-storybook` and see that it works

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This pull request has been released as version [`0.0.0-pr-24676-sha-214a6f84`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-24676-sha-214a6f84). Install it by pinning all your Storybook dependencies to that version.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-24676-sha-214a6f84`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-24676-sha-214a6f84) |
| **Triggered by** | @JReinhold |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`experiment-addon-peer-dep`](https://github.com/storybookjs/storybook/tree/experiment-addon-peer-dep) |
| **Commit** | [`214a6f84`](https://github.com/storybookjs/storybook/commit/214a6f845fcb3af4e07839ff9288c71f5d6a0e02) |
| **Datetime** | Wed Nov  8 11:20:01 UTC 2023 (`1699442401`) |
| **Workflow run** | [6797494862](https://github.com/storybookjs/storybook/actions/runs/6797494862) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=24676`_
</details>

### CODE REVIEW

**Comment by JReinhold** [comment]
@shilman could I get you to run benchmarks for this branch/version? It's largely complete now.

We expect unchanged performance and build size, but we expect a noticeable reduction in install size.

File: code/addons/controls/package.json [hunk 1]
```diff
@@ -32,31 +32,13 @@
...
-      "require": "./dist/manager.js",
-      "import": "./dist/manager.mjs"
-    },
-    "./register": {
-      "types": "./dist/manager.d.ts",
-      "require": "./dist/manager.js",
-      "import": "./dist/manager.mjs"
-    },
+    "./manager": "./dist/manager.js",
+    "./register": "./dist/manager.js",
```
Comment (JReinhold) [inline code/addons/controls/package.json:36]: manager and preview entries don't need types nor cjs outputs as they are always loaded in the manager or preview in the browser, never directly imported by the user.

File: code/addons/controls/package.json [hunk 2]
```diff
@@ -66,43 +48,35 @@
...
-    "entries": [
-      "./src/index.ts",
-      "./src/manager.tsx"
+    "exportEntries": [
+      "./src/index.ts"
     ],
-    "platform": "browser"
+    "managerEntries": [
+      "./src/manager.tsx"
+    ]
```
Comment (JReinhold) [inline code/addons/controls/package.json:79]: these new fields are used by `addon-bundle.ts` to determine which entry points needs which packages externalized.

File: code/addons/docs/src/preset.ts [hunk 3]
```diff
@@ -194,4 +194,12 @@ const docsX = docs as any;
...
 ensureReactPeerDeps();
 
-export { webpackX as webpack, indexersX as experimental_indexers, docsX as docs };
+const optimizeDeps = [
+  '@mdx-js/react',
+  '@storybook/addon-docs > acorn-jsx',
+  '@storybook/addon-docs',
+  '@storybook/addon-essentials/docs/mdx-react-shim',
+  'markdown-to-jsx',
+];
```
Comment (JReinhold) [inline code/addons/docs/src/preset.ts:203]: Moved from the Vite builder's default optimization, as that would break if `addon-essentials` was installed without `react`, which is a valid scenario now.

File: code/addons/docs/src/preset.ts [hunk 4]
```diff
@@ -194,4 +194,12 @@ const docsX = docs as any;
 
 ensureReactPeerDeps();
 
-export { webpackX as webpack, indexersX as experimental_indexers, docsX as docs };
+const optimizeDeps = [
```
Comment (JReinhold) [inline code/addons/docs/src/preset.ts:197]: @ndelangen I think we should move this to `framework.options.builder.optimizeDeps`, where other Vite-specific options are set too.

Comment (ndelangen) [reply code/addons/docs/src/preset.ts:197]: why? that just overly complicates things, no?

**Comment by ndelangen** [comment]
CI is failing over this:
https://github.com/storybookjs/storybook/blob/960426dba363fe7b7b3a4b487aed5e057ac1a7f5/code/renderers/preact/template/cli/Button.jsx#L1

Why does our example preact component have a need for `prop-types`?
If if does indeed need it, then we should ensure preact users actually get that dependency injected into their package.json via `sb init``

File: code/addons/docs/src/preset.ts (hunk 4 above)
Comment (JReinhold) [reply code/addons/docs/src/preset.ts:197]: It just feels natural to locate this very Vite-specific option with the other Vite-specific options. From an API-design mindset I don't see why we would make it top-level, unless it was because we didn't intend this to be a public API.

**Comment by JReinhold** [comment]
@ndelangen I've taken the liberty to refactor the globals logic in `manager` and `preview`, just renaming variables and moving files, but the same code flow. Doing this made it more clear to me what was going on, and it made me realize that `preview` should not have the same structure as `manager`, as Vite+Webpack doesn't use that same externalization plugin, so there's no need to generate that map of exports at all for `preview`.

File: code/lib/preview/src/globals/definitions.ts
Comment (ndelangen) [inline code/lib/preview/src/globals/definitions.ts:1]: Why is this removed again, didn't we add it for a good reason?

FYI, @valentinpalkovic will need this in his work/spike project.

File: code/addons/docs/src/preset.ts (hunk 4 above)
Comment (JReinhold) [reply code/addons/docs/src/preset.ts]: We decided to keep this API undocumented, and so we can make it a bit less "correct", and it is now just a top-level `optimizeViteDeps` config property.

File: code/lib/preview/src/globals/definitions.ts
Comment (JReinhold) [reply code/lib/preview/src/globals/definitions.ts:1]: @valentinpalkovic agrees that we can add this back in as part of his work if it will ever be needed.

**Comment by JReinhold** [comment]
@shilman benchmark comparisons between `next` as a baseline (top) and this branch (bottom):

I don't know how much of this is noise and how much is actual change.

## Vite

headers
next
this branch

## Webpack

headers
next
this branch

File: code/addons/a11y/package.json [hunk 5]
```diff
@@ -61,47 +49,40 @@
...
     "@storybook/types": "workspace:*",
-    "axe-core": "^4.2.0",
-    "lodash": "^4.17.21",
-    "react-resize-detector": "^7.1.2"
-  },
-  "devDependencies": {
     "@testing-library/react": "^11.2.2",
+    "lodash": "^4.17.21",
+    "react": "^16.8.0",
+    "react-dom": "^16.8.0",
```
Comment (ndelangen) [inline code/addons/a11y/package.json:70]: Did you check if the `react-dom` dep is needed truly in these addons?

It's not a big deal at all having them as devDependencies of course, no harm to users.

Comment (JReinhold) [reply code/addons/a11y/package.json:70]: no, I added it as a todo to my cooldown period.

> - [ ]  Remove `react-dom` from `package.json`s when not needed
> - [ ]  Audit dependencies of frameworks+renderers, they might have a lot of unused `@storybook` dependencies we can remove

**Review by shilman — APPROVED** [review]
I only reviewed for perf. I found that the sizes were slightly worse across the board, but not by an amount large enough to block the work. It would be useful to explain why the sizes all increased by a small amount, though, since it's not intuitive to me based on the change.

**Review by ndelangen — APPROVED** [review]

File: code/addons/controls/package.json (hunk 1 above)
Comment (StaphSynth) [reply code/addons/controls/package.json:36]: That's not true. My project was importing these types to allow us to overlay our own defaults. This change broke our workflow, it modified the exported output of the package and should not have been released under a minor tag as it changed the API.
